### PR TITLE
perf: Avoid looking up unnecessary TSDB symbols during Volume API

### PR DESCRIPTION
**Pull request by cyriltovena**
**What this PR does / why we need it**:

This avoid loading all symbols for stats queries when not needed.

```
❯ benchstat before.txt after.txt
name                 old time/op    new time/op    delta
TSDBIndex_Volume-16    1.10ms ± 2%    0.40ms ± 4%  -63.24%  (p=0.008 n=5+5)

name                 old alloc/op   new alloc/op   delta
TSDBIndex_Volume-16    84.8kB ± 0%    52.7kB ± 0%     ~     (p=0.079 n=4+5)

name                 old allocs/op  new allocs/op  delta
TSDBIndex_Volume-16     5.07k ± 0%     1.07k ± 0%  -78.95%  (p=0.008 n=5+5)

````
**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/stores/shipper/indexshipper/tsdb/single_file_index.go [hunk 1]
```diff
@@ -362,6 +359,24 @@ func (i *TSDBIndex) Volume(
 	seriesLabels := labels.Labels(make([]labels.Label, 0, len(labelsToMatch)))
 
 	aggregateBySeries := seriesvolume.AggregateBySeries(aggregateBy) || aggregateBy == ""
+	var by map[string]struct{}
+	if !includeAll && (aggregateBySeries || len(targetLabels) > 0) {
+		by = make(map[string]struct{}, len(labelsToMatch))
+		for k := range labelsToMatch {
+			by[k] = struct{}{}
+		}
+		if len(labelsToMatch) > 0 {
```
Comment (trevorwhitney) [inline pkg/storage/stores/shipper/indexshipper/tsdb/single_file_index.go:368]: why are we iterating over `labelsToMatch` twice?

Comment (cyriltovena) [reply pkg/storage/stores/shipper/indexshipper/tsdb/single_file_index.go:368]: Yeah this was supposed to be targetLabels

File: pkg/storage/stores/shipper/indexshipper/tsdb/single_file_index.go [hunk 2]
```diff
@@ -365,11 +368,10 @@ func (i *TSDBIndex) Volume(
 		for k := range labelsToMatch {
 			by[k] = struct{}{}
 		}
-		if len(labelsToMatch) > 0 {
-			for k := range labelsToMatch {
-				by[k] = struct{}{}
-			}
+		for _, k := range targetLabels {
+			by[k] = struct{}{}
```
Comment (trevorwhitney) [inline pkg/storage/stores/shipper/indexshipper/tsdb/single_file_index.go:372]: this should be the same as `labelsToMatch`. see https://github.com/grafana/loki/blob/9eb557cb3757e80083d47ad493629775722a2d9b/pkg/util/series_volume.go#L34

**Review by sandeepsukhani — COMMENTED** [review]
The changes look good to me. Once this is marked as ready for review, I will do a final review.

**Review by trevorwhitney — APPROVED** [review]
Looks good, need to test in GEL with LBAC before merging.

**Pull request merged**